hkmc2/shared/src/test/mlscript-compile/LazyFingerTree.mls

```diff
@@ -0,0 +1,165 @@
+    Splice(bits, l) as s then s.materialize().slice(beg, fin)
+    ft and
+      FingerTreeList.isFingerTree(ft) 
+        then FingerTreeList.slice(beg, (len - fin))(ft)
+        else View of arr, beg, (len - fin)
+
+// Marker base class for View and Splice
+class LazyFT extends IterableBase with
+
+  fun toString() = "[" + this joined(", ") + "]"
+  fun concat(...args) = concatHelper(mut [SpliceMarker, this], this.length, ...args)
+
+class View(val underlying: IterableBase, val start, val end) extends LazyFT with
+
+  fun iterator() =
+    let idx = start
+    let until = underlying.length - end
+    Iterator of () =>
+      if idx < until then
+        let next = underlying.at(idx)
+        set idx = idx + 1
+        Result.Next of next
+      else Result.Done
+
+  fun materialize() = underlying.slice(start, underlying.length - end)
+  
+  fun at(i) =
+    if i < -length || i >= length do
+      throw RangeError("View.at: Index out of bounds")
+    underlying.at(normIdx(i, length) + start)
+  fun slice(beg, fin) = sliceHelper(beg, fin, this)
+
+  val len = underlying.length - start - end
+  fun length = len
+  fun toString() = "[" + this joined(", ") + "]"
+
+object SpliceMarker
+class Splice(val bits, val len) extends LazyFT with
+  let materialized = null
+
+  fun iterator() = materialize().iterator()
+
+  fun reify =
+    let acc = FingerTreeList.mk()
+    let stack = mut []
+    stack.push([0, bits])
+    while stack.length > 0 do
+      let e = stack.pop()
+      let idx = e.[0]
+      let vals = e.[1]
+
+      if vals is Splice do
+        set vals = vals.bits
+
+      while (idx < vals.length and
+          (vals.at(idx) is SpliceMarker) is false) do
+        set acc = acc :+ vals.at(idx)
+        set idx += 1
+
+      if (idx < vals.length and vals.at(idx) is SpliceMarker) do
+        stack.push([idx + 2, vals])
+        let next = vals.at(idx + 1)
+        if FingerTreeList.isFingerTree(next)
+          then set acc = acc ++ next
+          else stack.push([0, next])
+    acc
+
+  fun materialize() =
+    if materialized == null do
+      set materialized = reify
+    materialized
+
+  fun at(i) = materialize().at(i)
+  // called on eager spreads in matches
+  fun slice(beg, fin) = sliceHelper(beg, fin, this)
+  fun concat(...args) =
+    if materialized === null 
+    then concatHelper(mut [SpliceMarker, this], len, ...args)
+    else concatHelper(mut [SpliceMarker, materialized], len, ...args)   
+  
+  fun length = len
+  fun toString() = "[" + materialize() joined(", ") + "]"
+
+
+module LazyFingerTree with...
+
+
+fun mk(...args) = View of FingerTreeList.mk(...args), 0, 0
+fun concat(...args) = concatHelper(mut [], 0, ...args)
+fun slice(beg, fin)(xs) = 
+  if beg < 0 || fin < 0 do
+    throw RangeError("LazyFingerTree.slice: indices must be non-negative")
+  if beg > xs.length || fin > xs.length do
+    throw RangeError("LazyFingerTree.slice: indices out of bounds")
+  sliceHelper(beg, xs.length - fin, xs)
```

Comment:
There is another slice method that takes in argument differently here as well.

Reply:
This should get rename the same as the other one.